compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -1280,20 +1315,77 @@ object Parsers {
       if (in.token == RPAREN) Nil else commaSeparated(exprInParens)
 
     /** ParArgumentExprs ::= `(' [ExprsInParens] `)'
-     *                    |  `(' [ExprsInParens `,'] PostfixExpr `:' `_' `*' ')' \
-     */
-    def parArgumentExprs(): List[Tree] =
-      inParens(if (in.token == RPAREN) Nil else commaSeparated(argumentExpr))
+     *                    |  `(' [ExprsInParens `,'] PostfixExpr `:' `_' `*' ')'
+     *
+     *  Special treatment for arguments of primary class constructor
+     *  annotations. Empty argument lists `(` `)` get represented
+     *  as `List(ParamNotArg)` instead of `Nil`, indicating that the
+     *  token sequence should be interpreted as an empty parameter clause
+     *  instead. `ParamNotArg` can also be produced when parsing the first
+     *  argument (see `classConstrAnnotExpr`).
+     *
+     *  The method affects `lookaheadTokens` as a side effect.
+     *  If the argument list parses as `List(ParamNotArg)`, `lookaheadTokens`
+     *  contains the tokens that need to be replayed to parse the parameter clause.
+     *  Otherwise, `lookaheadTokens` is empty.
+     */
+    def parArgumentExprs(): List[Tree] = {
+      if (inClassConstrAnnots) {
+        assert(lookaheadTokens.isEmpty)
+        saveLookahead()
+        accept(LPAREN)
+        val args =
+          if (in.token == RPAREN) ParamNotArg :: Nil
+          else {
+            openParens.change(LPAREN, +1)
+            try commaSeparated(argumentExpr)
+            finally openParens.change(LPAREN, -1)
+          }
+        if (args == ParamNotArg :: Nil) in.adjustSepRegions(RPAREN) // simulate `)` without requiring it
+        else {
+          lookaheadTokens.clear()
+          accept(RPAREN)
+        }
+        args
+      }
+      else
+        inParens(if (in.token == RPAREN) Nil else commaSeparated(argumentExpr))
+
+    }
 
     /** ArgumentExprs ::= ParArgumentExprs
      *                 |  [nl] BlockExpr
      */
     def argumentExprs(): List[Tree] =
       if (in.token == LBRACE) blockExpr() :: Nil else parArgumentExprs()
 
-    val argumentExpr = () => exprInParens() match {
-      case a @ Assign(Ident(id), rhs) => cpy.NamedArg(a)(id, rhs)
-      case e => e
+    val argumentExpr = () => {
+      val arg =
+        if (inClassConstrAnnots && lookaheadTokens.nonEmpty) classConstrAnnotExpr()
+        else exprInParens()
+      arg match {
+        case arg @ Assign(Ident(id), rhs) => cpy.NamedArg(arg)(id, rhs)
+        case arg => arg
+      }
+    }
+
+    /** Handle first argument of an argument list to an annotation of
+     *  a primary class constructor. If the current token either cannot
+     *  start an expression or is an identifier and is followed by `:`,
```

Comment:
Indeed, it seems that scalac always expects a primary constructor annotation to have exactly one parameter list, which is a more annoying restriction. I'm still not super enthusiastic about adding so much special-casing in the parser (this will need to be replicated in scala.meta for example, /cc @xeno-by), but I'm not opposed to it. The main annoyance is that if seemingly valid code is misinterpreted, you'll get a very confusing error message, but that could be left as a future improvement.